### ISSUE TITLE
mbedtls_time_t does not name a type in ssl.h

### ISSUE THREAD

**Issue by janjongboom** [ARMmbed/mbedtls#522]
Trying to compile a program that depends on mbed TLS on a Raspberry Pi Zero...

```
[98/126] Building CXX object ym/mbed-client-mbedtls/source/CMakeFiles/mbed-client-mbedtl...mbed-client-gateway/yotta_modules/mbed-client-mbedtls/source/m2mconnectionsecurity.cpp.o
FAILED: /usr/bin/c++   -DYOTTA_MODULE_NAME=mbed-client-mbedtls -std=gnu++11 -O2 -g -DNDEBUG -I/home/pi/connector-loriot/mbed-client-gateway -I/home/pi/connector-loriot/mbed-client-gateway/yotta_modules/mbed-client -I/home/pi/connector-loriot/mbed-client-gateway/yotta_modules/mbed-client-c -I/home/pi/connector-loriot/mbed-client-gateway/yotta_modules/mbed-trace -I/home/pi/connector-loriot/mbed-client-gateway/yotta_modules/mbed-client-linux -I/home/pi/connector-loriot/mbed-client-gateway/yotta_modules/mbed-client-mbedtls -I/home/pi/connector-loriot/mbed-client-gateway/yotta_modules/nanostack-libservice -I/home/pi/connector-loriot/mbed-client-gateway/yotta_modules/mbedtls -I/home/pi/connector-loriot/mbed-client-gateway/yotta_modules/mbed-client-mbedtls/source    -include "/home/pi/connector-loriot/mbed-client-gateway/build/arm-linux-native/yotta_config.h" -MMD -MT ym/mbed-client-mbedtls/source/CMakeFiles/mbed-client-mbedtls.dir/home/pi/connector-loriot/mbed-client-gateway/yotta_modules/mbed-client-mbedtls/source/m2mconnectionsecurity.cpp.o -MF ym/mbed-client-mbedtls/source/CMakeFiles/mbed-client-mbedtls.dir/home/pi/connector-loriot/mbed-client-gateway/yotta_modules/mbed-client-mbedtls/source/m2mconnectionsecurity.cpp.o.d -o ym/mbed-client-mbedtls/source/CMakeFiles/mbed-client-mbedtls.dir/home/pi/connector-loriot/mbed-client-gateway/yotta_modules/mbed-client-mbedtls/source/m2mconnectionsecurity.cpp.o -c /home/pi/connector-loriot/mbed-client-gateway/yotta_modules/mbed-client-mbedtls/source/m2mconnectionsecurity.cpp
In file included from /home/pi/connector-loriot/mbed-client-gateway/yotta_modules/mbedtls/mbedtls/debug.h:32:0,
                 from /home/pi/connector-loriot/mbed-client-gateway/yotta_modules/mbed-client-mbedtls/mbed-client-mbedtls/m2mconnectionsecuritypimpl.h:26,
                 from /home/pi/connector-loriot/mbed-client-gateway/yotta_modules/mbed-client-mbedtls/source/m2mconnectionsecurity.cpp:22:
/home/pi/connector-loriot/mbed-client-gateway/yotta_modules/mbedtls/mbedtls/ssl.h:545:5: error: ‘mbedtls_time_t’ does not name a type
     mbedtls_time_t start;       /*!< starting time      */
```

If I change the `#define` to read (in include/mbedtls/ssl.h):

```cpp
#if defined(MBEDTLS_HAVE_TIME)
#include <time.h>
#define mbedtls_time_t   time_t    # <-- this is new
#endif
```

It does compile. Ain't that interesting... Wonder why this is not triggered in test suite.

cc @pjbakker

**Comment by ciarmcom** [comment]
ARM Internal Ref: IOTSSL-841

**Comment by anttiylitokola** [comment]
@janjongboom, can you try to compile again? mbed-client-mbed-tls module is now updated to yotta containing following fix:
https://github.com/ARMmbed/mbed-client-mbed-tls/commit/4e81d7ff7811876f2b44613d1f88fb39c7e14781

**Comment by jcowgill** [comment]
FYI since this bug broke some applications in Debian, I have applied the following patch to the Debian packaging for the 2.3.0 release:
```diff
--- a/include/mbedtls/ssl.h
+++ b/include/mbedtls/ssl.h
@@ -52,7 +52,7 @@
 #endif
 
 #if defined(MBEDTLS_HAVE_TIME)
-#include <time.h>
+#include "platform.h"
 #endif
 
 /*
```

**Comment by janjongboom** [comment]
@anttiylitokola Yeah, that would do. But @jcowgill's comment makes me think this should be patched in mbedtls repo as well...

**Comment by pjbakker** [comment]
@janjongboom agreed. We'll pick it up.

**Comment by sithglan** [comment]
Hello,
I hit the same issue with curl and mbedtls 2.3.0. The above patch of @jcowgill fixed it for me. See also: http://permalink.gmane.org/gmane.comp.web.curl.library/47856 Please apply.

Cheers,
    Thomas

**Comment by sbutcher-arm** [comment]
This has now been resolved in a series of commits (b5b6af2, b928343 and 23e9778). Thanks to @jcowgill for the pointer, but the fix above isn't sufficient, and there was an underlying issue that `mbedtls_time_t` was dependent on the symbol `MBEDTLS_HAVE_TIME` which wasn't being properly respected.

The issue has now been reviewed, committed, tested and fixed in the above commits.
